indsl/filter/hilbert_huang_transform.py

```diff
@@ -0,0 +1,458 @@
+    the function returns the trend component itself. This allows users to analyze and study the trend separately
+    from the oscillatory components, providing insights into the underlying patterns and dynamics of the signal.
+
+    Args:
+        signal (pd.Series): Input time series signal.
+        sift_thresh (float, optional): The threshold used for sifting convergence. Defaults to 1e-8.
+        max_num_imfs (int, optional): The maximum number of oscillatory components to extract. If None, extract all IMFs. Defaults to None.
+        error_tolerance (float, optional): The tolerance for determining significant IMFs. Defaults to 0.05.
+        return_trend (bool, optional): Flag indicating whether to return the trend component of the signal. Defaults to True.
+
+    Returns:
+        pd.Series: The detrended signal if `return_trend` is False, otherwise, the trend component of the signal.
+
+    Raises:
+        ValueError: If `sift_thresh` is not a number higher than zero.
+        ValueError: If `max_num_imfs` is not an integer higher than zero.
+        ValueError: If `error_tolerance` is not higher than zero.
+        Any exceptions that may occur during Empirical Mode Decomposition (EMD).
+
+
+    References:
+        - Huang, Norden E., et al. "The empirical mode decomposition and the Hilbert spectrum for nonlinear and non-stationary time series analysis."
+        Proceedings of the Royal Society of London. Series A: Mathematical, Physical and Engineering Sciences 454.1971 (1998): 903-995.
+        - Yang, Zhijing & Bingham, Chris & Ling, Bingo & Gallimore, Michael & Stewart, Paul & Zhang, Yu. (2012).
+        "Trend extraction based on Hilbert-Huang transform." 1-5. 10.1109/CSNDSP.2012.6292713.
+    """
+    from PyEMD import EMD
+
+    if sift_thresh <= 0:
+        raise ValueError("The sifting threshold must be a number higher than zero")
+    if max_num_imfs is not None and max_num_imfs <= 0:
+        raise ValueError("The maximum number of oscillatory components must be an integer higher than zero")
+    if error_tolerance <= 0:
+        raise ValueError("The energy tolerance must be higher than zero")
+
+    signal_array = signal.to_numpy()
+
+    emd = EMD()
+    imfs = emd(signal_array)  # Empirical Mode Decomposition
+
+    # Truncate the IMFs to the max number of IMFs
+    if max_num_imfs is not None and max_num_imfs <= len(imfs):
+        imfs = imfs[:max_num_imfs]
+
+    # Compute the analytic signal for each IMF using the Hilbert transform and store them in a list
+    analytic_signals = []
+    for imf in imfs:
+        analytic_signal = hilbert(imf)
+        analytic_signals.append(analytic_signal)
+
+    # Find the total number of IMFs and the index of the last IMF
+    total_number_of_imf: int = imfs.shape[1]
+    index_of_the_last_imf: int = total_number_of_imf - 1
+    significant_imf_index_rho = len(imfs) - 1
```

Comment:
It is possible to merge these three lines into one as only the significant_imf_index_rho is used